### ISSUE TITLE
False positive on "__imul__" and "__mul__" signatures

### ISSUE THREAD

**Issue by mttbernardini** [python/mypy#15787]
**Bug Report**

Defining `__mul__` and `__imul__` with `@overload` behaviour yields a `error: Signatures of "__imul__" and "__mul__" are incompatible`, although the `@overload` declarations match.

**To Reproduce**

```python
from typing import TypeVar, Self, overload

class Vec:
    @overload
    def __mul__(self, other: float) -> Self:
        """Performs a vector-by-scalar multiplication"""
    @overload
    def __mul__(self, other: Vec) -> float:
        """Performs a dot-product between vectors"""
    
    def __mul__(self, other):
        if isinstance(other, float):
            return type(self)()
        elif isinstance(other, Vec):
            return 0.0
        else:
            return NotImplemented
    
    
    @overload
    def __imul__(self, other: float) -> Self:
        """Performs a vector-by-scalar multiplication in-place"""
    @overload
    def __imul__(self, other: Vec) -> float:
        """Performs a dot-product between vectors in-place"""
    
    def __imul__(self, other):
        if isinstance(other, float):
            return self
        elif isinstance(other, Vec):
            return 0.0
        else:
            return NotImplemented
```

[(mypy playground link)](https://mypy-play.net/?mypy=master&python=3.11&flags=check-untyped-defs&gist=21452685ffb865bbb2172659cb2174c1)

**Expected Behavior**

No errors

**Actual Behavior**

```
main.py:27: error: Signatures of "__imul__" and "__mul__" are incompatible  [misc]
```

**Your Environment**

- Mypy version used: master branch
- Mypy command-line flags: `--check-untyped-defs`
- Python version used: 3.11

**Comment by tyralla** [comment]
duplicate of #14956
